### PR TITLE
[HOLD] Add stack counter for command port closures

### PR DESCRIPTION
**Pull request by coleaeason**
@tylerkaraszewski try two, this one has a mutex around opening a command port to prevent two threads for both trying to open one at the same time. Ran the auth tests several times with no failures, before all billing tests failed. Ran the bedrock tests with no failures.

### CODE REVIEW

File: BedrockServer.cpp [hunk 1]
```diff
@@ -1187,8 +1188,8 @@ void BedrockServer::postPoll(fd_map& fdm, uint64_t& nextActivity) {
     }
     if (!_suppressCommandPort && (state == SQLiteNode::MASTERING || state == SQLiteNode::SLAVING) &&
         _shutdownState.load() == RUNNING) {
-        // Open the port
-        if (!_commandPort) {
+        // Open the port if we don't have one and we are the only thread trying to open it.
+        if (!_commandPort && _commandPortMutex.try_lock()) {
```
Comment (tylerkaraszewski) [inline BedrockServer.cpp]: That's not how `try_lock` works. create a `lock_guard` around the block of code that does the open.

File: BedrockServer.cpp [hunk 2]
```diff
@@ -1530,6 +1531,15 @@ void BedrockServer::suppressCommandPort(const string& reason, bool suppress, boo
     } else {
         // Clearing past suppression, but don't reopen (It's always safe to close, but not always safe to open).
         SHMMM("Clearing command port suppression");
+        closures = _commandPortClosures.fetch_sub(1);
+        if (closures == 1 && _commandPortMutex.try_lock()) {
```
Comment (tylerkaraszewski) [inline BedrockServer.cpp]: Same here, just create a `lock_guard` as the first line in this function.

File: BedrockServer.cpp [hunk 3]
```diff
@@ -1187,8 +1189,8 @@ void BedrockServer::postPoll(fd_map& fdm, uint64_t& nextActivity) {
     }
     if (!_suppressCommandPort && (state == SQLiteNode::MASTERING || state == SQLiteNode::SLAVING) &&
         _shutdownState.load() == RUNNING) {
-        // Open the port
-        if (!_commandPort) {
+        // Open the port if we don't have one and we've had as many opens as we have closures.
+        if (!_commandPort && _commandPortMutex.try_lock()) {
```
Comment (tylerkaraszewski) [inline BedrockServer.cpp]: That's not really how `try_lock` works, create a `lock_guard` around the block of code that calls `openPort`.

File: BedrockServer.cpp [hunk 4]
```diff
@@ -1519,17 +1525,29 @@ void BedrockServer::suppressCommandPort(const string& reason, bool suppress, boo
...
             _portPluginMap.clear();
             _commandPort = nullptr;
         }
     } else {
         // Clearing past suppression, but don't reopen (It's always safe to close, but not always safe to open).
         SHMMM("Clearing command port suppression");
+        lock_guard<decltype(_commandPortMutex)> lock(_commandPortMutex);
+        closures = _commandPortClosures.fetch_sub(1);
+        if (closures == 1 && !_commandPort) {
+            SINFO("Removing a command port closure.");
```
Comment (tylerkaraszewski) [inline BedrockServer.cpp]: Weird log line in this case.

**Comment by coleaeason** [comment]
Closing this, the problem has gone away since I made this. if i comes back I'll just revisit.

**Pull request closed without merging**